packages/server/modules/core/tests/favoriteStreams.spec.js

```diff
@@ -0,0 +1,294 @@
+ */
+
+/**
+ * Cleaning up relevant tables
+ */
+async function cleanup() {
+  await truncateTables([StreamFavorites.name, Streams.name, Users.name])
+}
+
+const favoriteMutationGql = gql`
+  mutation ($sid: String!, $favorited: Boolean!) {
+    streamFavorite(streamId: $sid, favorited: $favorited) {
+      id
+      favoritedDate
+      favoritesCount
+    }
+  }
+`
+
+const favoriteStreamsQueryGql = gql`
+  query ($cursor: String, $limit: Int! = 10) {
+    user {
+      id
+      favoriteStreams(cursor: $cursor, limit: $limit) {
+        totalCount
+        cursor
+        items {
+          id
+        }
+      }
+    }
+  }
+`
+
+describe('Favorite streams', () => {
+  const myPubStream = {
+    name: 'My Stream 1',
+    isPublic: false
+  }
+  const myStream = {
+    name: 'My Stream 2',
+    isPublic: true
+  }
+  const notMyStream = {
+    name: 'Not My Stream 1',
+    isPublic: false
+  }
+  const notMyPubStream = {
+    name: 'Not My Stream 2',
+    isPublic: true
+  }
+  let me = {
+    name: 'Itsa Me',
+    email: 'me@gmail.com',
+    password: 'sn3aky-1337-b1m'
+  }
+  let otherGuy = {
+    name: 'Some Other DUde',
+    email: 'otherguy@gmail.com',
+    password: 'sn3aky-1337-b1m'
+  }
+
+  before(async function () {
+    await cleanup()
+
+    // Seeding
+    await Promise.all([
+      createUser(me).then((id) => (me.id = id)),
+      createUser(otherGuy).then((id) => (otherGuy.id = id))
+    ])
+
+    await Promise.all([
+      createStream({ ...myPubStream, ownerId: me.id }).then((id) => (myPubStream.id = id)),
+      createStream({ ...myStream, ownerId: me.id }).then((id) => (myStream.id = id)),
+      createStream({ ...notMyStream, ownerId: otherGuy.id }).then((id) => (notMyStream.id = id)),
+      createStream({ ...notMyPubStream, ownerId: otherGuy.id }).then(
+        (id) => (notMyPubStream.id = id)
+      )
+    ])
+  })
+
+  after(async () => {
+    await cleanup()
+  })
+
+  describe('when authenticated', () => {
+    /** @type {import('apollo-server-express').ApolloServer} */
+    let apollo
+
+    const favoriteStream = async (sid, favorited) =>
+      await apollo.executeOperation({
+        query: favoriteMutationGql,
+        variables: { sid, favorited }
+      })
+
+    before(async () => {
+      apollo = buildApolloServer({
+        context: () =>
+          addLoadersToCtx({
+            auth: true,
+            userId: me.id,
+            role: Roles.Server.User,
+            token: 'asd',
+            scopes: AllScopes
+          })
+      })
+
+      // Drop all favorites to ensure we don't favorite already favorited streams
+      await StreamFavorites.knex().truncate()
+    })
+
+    const accessibleStreamIds = [
+      [() => myPubStream.id, 'owned and public'],
+      [() => myStream.id, 'owned and not public'],
+      [() => notMyPubStream.id, 'not owened, but public']
+    ]
+
+    accessibleStreamIds.forEach(([id, msgSuffix]) => {
```

Comment:
apparently this is how dataProviders are implemented in mocha

Reply:
ok, i can get behind this. Even if the syntaxt is a bit weird to me atleast.

Reply:
yea in jest there are built-in helpers for this, like `it.each(data)(message, (a,b,c) => {..});`, but in mocha there's nothing like that and you just use a basic array forEach for this